### ISSUE TITLE
Support negative fees

### ISSUE THREAD

**Issue by mikejolley** [woocommerce/woocommerce#16528]
After our venture into discounts, it looks like supporting negative fees will be the simpler solution due to our technical debt.

- [x] Split fees API out of cart into own class
- [x] Allow negative fees
- [x] Calculates taxes correctly when using a negative fee
- [x] Show in admin correctly
- [x] Update tests

**Comment by roberttrigg** [comment]
Will the Free Shipping minimum order amount take in to account negative fees? At the moment, you can apply a negative fee as a discount, but Free Shipping will still be available even though the fee takes the order total below the Free Shipping threshold.

**Comment by mikejolley** [comment]
Fees are calculated after shipping, so it shouldn't affect that.

**Comment by roberttrigg** [comment]
If negative fees are going to be used for discounts, won't they will need to be calculated before shipping? At the moment this behaviour is not what you would expect as you wouldn't want Free Shipping to be available if the final total is below the threshold.

Example in the screenshot:

![screen shot 2017-08-22 at 18 30 40](https://user-images.githubusercontent.com/25773113/29578833-95c93c2c-8768-11e7-89ea-7a26600c1cbb.png)

- subtotal qualifies for free shipping (£30 minimum)
- free shipping available.
- Negative fee applied as a discount
- the final total is now below the free shipping threshold

If this behaviour will remain with the supporting of negative fees, is it the behaviour we would need / expect?

**Comment by mikejolley** [comment]
Fees are after shipping. Negative fees can 'discount' shipping.

If you want it done before shipping, don't use them :) Use coupons.

**Comment by mikejolley** [comment]
^ that multibuy example doesn't really look suitable for fees FYI. It should be filtering item prices?

You can still create 'dynamic' coupons in WooCommerce if needed.

**Comment by roberttrigg** [comment]
Thanks, I appreciate the feedback.

What would be a good case or example to use negative amounts for a fee? What is the reason for accommodating negative fees going forwards?

**Comment by pmgarman** [comment]
Backwards compatibility to not break people who used the bug that we treated as a feature.

As of 3.2 it would be better to just use the discounts api that is coming.

**Comment by roberttrigg** [comment]
thanks. I'm looking forward to testing the new api.

**Comment by mikejolley** [comment]
@pmgarman We decided not the add the fixed/percent based discounts API - it added extra complexity and didn't really offer anything beyond the -ve fees.

Coupons should be more flexible though.

**Comment by psaikali** [comment]
Hey there,

I've updated WooCommerce to 3.2 and noticed a weird bug with using negative cart fees.
For some reason, taxes are added whereas I'm specifically telling the `add_fee()` method not to add tax.

Adding negative fee with...
`WC()->cart->add_fee(hr_get_var('commissions_fee_name'), -$discount, false, '');`

... the fee object looks like :
```stdClass Object
(
    [id] => utilisation-de-vos-haricots
    [name] => Utilisation de vos haricots
    [tax_class] => 
    [taxable] => 
    [amount] => -25.3
    [total] => -25.3
    [tax_data] => Array
        (
            [3] => -0.69575
            [1] => -2.53
        )

    [tax] => -3.23
)```

Any idea why ?
Thanks !

**Comment by soulstyle** [comment]
Hi!

I just notice the same bug as @psaikali. On a LIVE site!!!!

Need a fix to this urgently!


Thanks a lot for any type of insight!

**Comment by mikejolley** [comment]
Taxes are indeed added because you're essentially discounting. So the taxes are discounted accordingly.

**Comment by psaikali** [comment]
Thanks @mikejolley for answering. 
Just a friendly feedback as a developer : this tweak added taxes calculations on negative fees, whereas the ->add_fee() method clearly offers a $taxable parameter. That's a bit of a shame :/

What's the best plan for us to get back to previous behavior (I downgraded to 3.1.2 on a client website to avoid difference of calculations) ?

Thanks for your help.

**Comment by mikejolley** [comment]
When you're discounting it doesn't matter if the fee is taxable or not; you're changing existing totals. You cannot charge tax on something not paid.

If I had a $10 order with $2 tax and a -ve fee of $10, I shouldn't pay $2 tax on nothing.

**Comment by psaikali** [comment]
OK, thanks for clarification !

**Comment by Sheaffy** [comment]
@mikejolley  If you are no longer adding a Discounts API, perhaps add the feature to turn negative fees as a actual system recognised discount so that in reports the discounts total can be calculated?

**Comment by enriquejros** [comment]
@mikejolley In Spain we need in some cases to add deductions applied to the subtotal. Those deductions don't imply a change in the taxation, because they have be paid later to IRS, so the must be non-taxable, even when they are negative. So, yes, sometimes you have to charge a tax on non paid amounts.

**Comment by LukasSteindl** [comment]
Have a similar problem related to tax. 

My product prices already include shipping and i deduct the shipping cost when the customer selects
to pickup his items up in the store ("Selbstabholung"). 

![image](https://user-images.githubusercontent.com/15838914/37053721-37a8462e-217d-11e8-84f3-9285bf39cb46.png)

However the 13% VAT is still calculated on the interim result and not on the total amount (deduced by the negative shipping fees). 

Expected VAT is 131.32/113*13=15.10EUR instead of 167.90/113*13=19.32EUR.

**Comment by LukasSteindl** [comment]
here are more images to document the problem: 
for positive shipping cost things work well:
![beipositivenshippingkostenrichtig](https://user-images.githubusercontent.com/15838914/37060458-a9f8a1ca-2190-11e8-9fef-2af3dcf296db.jpg)

however negative shipping costs show no negative tax:
![trotzsteuerbarem-shipping-rechnung-falsch](https://user-images.githubusercontent.com/15838914/37060496-c4e88284-2190-11e8-8389-8cb6239a7f68.JPG)
while the config clearly indicates that it should consider VAT for this type of shipping:
![steuern der versandkosten sollen dem standard entsprechen](https://user-images.githubusercontent.com/15838914/37060534-e0e9079c-2190-11e8-8a84-e7bcb29e15df.JPG)
![trotzsteuerbarem shipping wird es nicht in der rechnung berucksichtigt](https://user-images.githubusercontent.com/15838914/37060539-e4ce11d6-2190-11e8-9961-1fbe66ae7902.JPG)

**Comment by LukasSteindl** [comment]
the current workaround is to manually adjust each negative shipping-cost position to 
corrected_shipping_vat =  computed_shippingcost/113*13 (where 13% is the tax rate) and then setting the computed_shipping to computed_shipping_cost - corrected_shipping_vat . 

After that we resend the another mail to the customer with the corrected invoice.

**Comment by tobiushirogeri** [comment]
What happens to products that are not taxable like food and other products.
If I add a negative fee for non taxable products, It adds up a negative tax amount.
            $woocommerce->cart->add_fee( 'Bonis Dollars - '.$v['name'], -$v['amount'],false);

![selection_081](https://user-images.githubusercontent.com/8570089/39269838-fc5d6490-48a1-11e8-8919-5fc9f436a956.png)

**Comment by argordmel** [comment]
Hello

I have the same problem as @psaikali and @soulstyle. How can apply a negative discount after total cart?

Thanks!

**Comment by jamesckemp** [comment]
I experienced the same issue here and wanted to log my solution for future me (and anyone else reading).

In my case, I was adding a discount based on a set of products. I.e. discount of 20% when X, Y, and Z are purchased together.

I was doing the calculation based on the product price, which was resulting in the incorrect discount being applied (20% discount, plus tax).

Instead, I needed to base the discount amount on the product prices *excluding tax* using `wc_get_price_excluding_tax()`. Now the calculations are correct in every tax scenario. I imagine the same applies to shipping rates.
